### ISSUE TITLE
Documentation

### ISSUE THREAD

**Issue by mansellan** [WaynePhillipsEA/twinbasic#308]
**Documentation**
There is currently no official language documentation for twinBASIC.

**Describe the solution you'd like**
A documentation site, ideally to include:
- API documentation
- Getting started
- Conceptual topics

**Describe alternatives you've considered**
As a new language, twinBASIC is helped by the fact that is is source-compatible with VBA and VB6. However, there are improvements (overloads, generics) which do not yet have documentation.

**Additional context**
Language documentation is a non-trivial task. Fortunately, twinBASIC is effectively a superset of VB6, which is a (slight) superset of VBA.

The documentation for VBA is community-maintained on GitHub, under a [Creative Commons Attribution 4.0 International Public License](https://github.com/MicrosoftDocs/VBA-Docs/blob/master/LICENSE). This permits redistribution and modification (with attribution), and could form the basis of twinBASIC documentation (caution, I'm not a lawyer!) If I understand correctly, the CCA4 license would have to be maintained on the derivative work (i.e. the twinBASIC docs) if this route were taken.

This being the case, the task of creating and maintaining twinBASIC documentation could be a community effort, with tB devs retaining editorial control (merge rights).

I'm sure there's lots to consider here, but wanted to kick off the discussion.

**Comment by dmrvb** [comment]
Can't really produce serious documentation till the code is more stable but a short beginners guide would be great, lots of info can only be found by wading through the posts here.
On a related note, a forum might be really good so we can help each other leaving Wayne to get on with the coding. This could be a dedicated forum, which has costs, but maybe for now it could be a new thread on VBforums?????

**Comment by WaynePhillipsEA** [comment]
I absolutely agree with this.  If the licence allows us to use the VBA docs as a base, then that would be fantastic, so we should look further into that.

Words are not really my thing, so my original plan was to outsource this part of the project.  That said, I am totally behind a community based approach if there's enough interested parties.

Assuming we do go with this, I guess hosting the docs on GitHub makes the most sense.

**Comment by WaynePhillipsEA** [comment]
@dmrvb I tried to reach out to the site admin at vbforums a few weeks ago (via the listed email address), to discuss creating a twinBASIC sub-forum.  Unfortunately I didn't get a response.   Though if anyone here has contacts over at vbforums I'd be keen to talk to them.

**Comment by WaynePhillipsEA** [comment]
Just trying to gauge how many here would be willing or interested in contributing to community driven documentation for twinBASiC.  Could you please +1 this comment if you'd be interested.

**Comment by dmrvb** [comment]
I might be able to help, I am good at proof reading technical documents and have a liking for clear plain English writing rather than too much computer science jargon.
However my lifestyle is such that I do not always have spare time, and occasionally don't even have internet access, so I should not be trusted with time critical tasks.

**Comment by WaynePhillipsEA** [comment]
@dmrvb that's absolutely fine!  Any help would be fantastic, no matter how small the contribution.  I was just wanting a rough idea to make sure it's not just a single person willing/wanting to help with this before moving the idea forward.

**Comment by mansellan** [comment]
I've been looking into the VBA-Docs repo. Unsurpisingly, much of it is not relevant as it relates to the object models of specific Office applications, or to aspects of the VBE. Howver, all language constructs are documented, mostly under here: https://docs.microsoft.com/en-us/office/vba/language/reference/user-interface-help/visual-basic-language-reference

The repo layout itself is pretty suboptimal, everything is kinda mixed together. I'm just going through to see if I can put together a more sensible folder structure locally.

Content in the repo is into HTML using [DocFX](https://dotnet.github.io/docfx/), however I believe parts of the build chain are not divulged - it's using the (not yet released) v3 of DocFx, and all the template and styling assets are missing. I also couldn't see how it is building its TOC.

**Comment by WaynePhillipsEA** [comment]
Thanks @mansellan, I had a brief look through the files yesterday and came to the same conclusion.  I'm not too concerned about the DocFX HTML building from the markdown files, as that should(?) be straightforward to recreate provided we can work out how the TOC is generated.

Good luck with the file/folder structure, as that looks horrible.

**Comment by EduardoVB** [comment]
> Howver, all language constructs are documented, mostly under here: https://docs.microsoft.com/en-us/office/vba/language/reference/user-interface-help/visual-basic-language-reference

[Here](https://web.archive.org/web/20150101183954/http://msdn.microsoft.com/en-us/library/aa338032(v=vs.60).aspx) there is an archived version of the VB6 documentation.

**Comment by mansellan** [comment]
@WaynePhillipsEA so it's pretty simple (albiet tedious) to find all the topics and rename / relocate them. Happy to finish that over the next few days and push it to a repo. After that, each topic will need to be manually reviewed to remove references to VBA and Office, and fix up any cross-refeences. The TOC is pretty simple to set up, assuming you're sticking with DocFX.

Thanks @EduardoVB. Microsoft actually have the VB6 docs online in their archive [here](https://docs.microsoft.com/en-us/previous-versions/visualstudio/visual-basic-6/visual-basic-6.0-documentation) (albeit with many broken links), the trouble is that they've never (AFAICS) made it available under a permissive license and still retain copyright. For that reason, I don't think tB can use any of it.

**Comment by EduardoVB** [comment]
> the trouble is that they've never (AFAICS) made it available under a permissive license and still retain copyright. For that reason, I don't think tB can use any of it.

Yes, I don't mean to copy it directly, but to know about it to read it if needed to get the idea of what should be covered in the explanations, and as a source of information for the own understanding.

And also because in the VBA documentation there are things that are not covered because VBA lacks of many things that VB6 has.
Probably not at the language core, but as soon as we go to the environment it lacks things like UserControls, PropertyPages, Printer, Screen (not sure about this one), PropertyBags? PictureBoxes? common controls, Sstab, Msflexgrid, etc.
